[Feature] (La)yer-Neigh(bor) sampling implementation

mfbalin opened the pull request:
## Description
CPU&GPU implementation of LABOR sampling, a drop-in replacement to Neighbor Sampling with same fanout hyperparameters. This work was done partly during a summer NVIDIA Devtech AI internship under @nv-dlasalle's mentorship.
arXiv paper can be found here: [arXiv:2210.13339](https://arxiv.org/abs/2210.13339)
To run, either replace any dgl.dataloading.NeighborSampler with dgl.dataloading.LaborSampler, or rebase #4718 onto this PR, and then:
`cd examples/pytorch/labor`
`python train_lightning_labor.py`, to compare with neighbor sampling, `python train_lightning_labor.py --sampler=neighbor`.
With 1000 batch size and 3 layers with 10,10,10 fanout on reddit, can sample 7x less vertices with same quality compared to neighbor sampling with `importance-sampling=-1` option, while `importance-sampling=0` option samples around 5x less and is faster than Neighbor Sampling in general.

Weighted sampling is available too, seems to work as expected.

![same_loss_zoom](https://user-images.githubusercontent.com/16190101/193340540-c1ede60f-1867-4461-80cd-c5c32cb392c3.svg)
The loss curve on different datasets with same batch size. The soft edges represent the confidence interval. Number of sampled vertices can be found in Table below:
![image](https://user-images.githubusercontent.com/16190101/193340731-297687ab-86c0-43ad-b6a1-b80b5cf96216.png)
![same_budget_ordered](https://user-images.githubusercontent.com/16190101/193340552-baab283f-5c07-4cd1-826d-5da8d919c740.svg)
Vertex sampling efficiency under the same sampling budget. A starting batch size of 1k is used and the batch size is adjusted at the end of each epoch to better match the vertex budget. The first row of plots is for the batch size and the second row of plots shows the number of vertices sampled in the last layer, $|V_3|$ of running averages. The last row shows the training loss curves.


## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes
<!-- You could use following template
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)
-->

mfbalin commented:
@dgl-bot

BarclayII commented:
@dgl-bot

BarclayII commented:
@mfbalin What is the difference between this LABOR algorithm and LADIES?  If there is any difference, did you document the algorithm of LABOR algorithm somewhere?

Also I guess `sample_labors` is probably not a good name because `labor` is actually an English word and it does not refer to a specific algorithm so far (like the word `Shadow` in `ShaDow-GNN`).  I would write out the full name of it (like `LayerwiseSampler` or something).

mfbalin commented:
> @mfbalin What is the difference between this LABOR algorithm and LADIES? If there is any difference, did you document the algorithm of LABOR algorithm somewhere?
> 
> Also I guess `sample_labors` is probably not a good name because `labor` is actually an English word and it does not refer to a specific algorithm so far (like the word `Shadow` in `ShaDow-GNN`). I would write out the full name of it (like `LayerwiseSampler` or something, or just `LADIESSampler` if there is no difference between LABOR and LADIES).

It is a new algorithm, just like neighbor sampling, in samples fanout neighbors for each vertex but does this in a layer sampling fashion. Since it is a new algorithm and is a combination of layer and neighbor sampling, I call it labor sampling. A paper explaining the proposed method is currently under submission.

mfbalin commented:
@BarclayII This PR contains only the core of LABOR sampling now. #4718 has the example. Should be ready for reviews.

yaox12 commented:
@dgl-bot

mfbalin commented:
Why is there a problem with the linter? The linter passes in my local machine. I also tried inside the following docker that the ci seems to use for the lint check: dgllib/dgl-ci-lint:latest

BarclayII commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,176 @@
+from .base import BlockSampler
+from ..random import choice, seed
+from .. import backend as F
+from collections import defaultdict
```
For code style conformation please refer to https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit

BarclayII commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,176 @@
+    labor sampling for multilayer GNN.
+
+    This sampler will make every node gather messages from a fixed number of neighbors
+    per edge type. The neighbors are picked uniformly with default parameters.
```
Please remember to update the docstring with the arXiv link once you post it.

BarclayII commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,176 @@
+    ...      ('user', 'plays', 'game'): 4,
+    ...      ('game', 'played-by', 'user'): 3}] * 3)
+
+    If you would like non-uniform neighbor sampling:
+
+    >>> g.edata['p'] = torch.rand(g.num_edges())   # any non-negative 1D vector works
+    >>> sampler = dgl.dataloading.NeighborSampler([5, 10, 15], prob='p')
```
Remove.

mfbalin replied:
Labor can do weighted sampling too, just fixed the documentation there.

BarclayII commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,176 @@
+    >>> sampler = dgl.dataloading.as_edge_prediction_sampler(sampler)
+    >>> dataloader = dgl.dataloading.DataLoader(
+    ...     g, train_eid, sampler,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
```
Did you test with `as_edge_prediction_sampler`?

mfbalin replied:
Yes, I did. It seems to work. I don't have tests written for those scenarios though.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,276 @@
+
+        If -1 is given for a single edge type, all the neighboring edges with that edge
+        type will be selected.
+    random_seed : tensor with 3 or 4 integers
+        The random seed(s) to be used for sampling along with batch dependency count information
+        stores as the last two elements
```
Why do we want to have an individual `random_seed` for LABOR sampler?

mfbalin replied:
Because when doing distributed sampling, if seeds are the same, then every machine rolls same random numbers as if everything is running on a single machine. So labor sampling can do distributed sampling with no communication.

mfbalin replied:
Cooperative minibatching PR makes use of this feature.

mfbalin replied:
One extra use case makes use of having 2 random seeds also.

BarclayII replied:
Hmm having this in the `sample_labor` function seems adding additional complexity to the interface.  Is it possible to set the seed outside this function?

mfbalin replied:
I store the seed and also the count information in the LaborSampler class and pass it into sample_labors. Where else do you think I can set it and make it so that the c++ side receives it?

BarclayII replied:
For instance, `dgl.random.seed()` can set the C++ seeds in DGL.

BarclayII replied:
I also don't understand why we need an array of seeds.

mfbalin replied:
2 random seeds are used when one wants to use batch_dependency, making consecutive batches dependent on each other by smoothly switching between 2 seeds over batch_dependency batches. The other 2 numbers keep track of how many batches were generated and the batch dependency information. If I use a different seed for each different batch, then the batches are fully independent. If I use the same seed for two different batches, then they are fully dependent. If I want to have something inbetween these two use cases, then I can pass 2 seeds along with how much I want to use the first seed and how much of the second seed. Then it will be a combination of the two seeds and something inbetween having fully independent and dependent batches.

This use case is not in the arxiv paper, it will explained in the cooperative minibatching paper. Also, I didn't want to depend on DGL's seed because if the user uses DGL PRNG for other uses, then it will change the state of the PRNG.

mfbalin replied:
Making batches dependent on each other has the advantage that the rolled randoms will be same for same vertices. Then more of the same vertices will be sampled. This reduces input feature requirements across multiple batches.

mfbalin replied:
If one uses a gpu cache in the #4341 to cache input features, one will observe that using a larger batch_dependency will improve the cache hit rate.

BarclayII replied:
Hmm if that's the case, I would move the `batch_dependency`-related code to the cooperative minibatching PR (or another PR dedicated for cooperative minibatching on LABOR sampling).

Speaking of the interface, I think the core idea is to have a more fine-grained RNG control on each operator.  I don't think supplying seeds in the operators is a good idea as (1) the RNG will then be coupled with the operator itself, (2) it adds extra complexity to the operator's signature, and (3) it conflicts with `dgl.random.seed` which can cause extra confusion to the users.  PyTorch for instance does fine-grained RNG control with context managers, like
```python
with torch.random.fork_rng():
    torch.random.set_rng_state(...)
    x = torch.randn(...)
```
We can borrow the same idea, where you can simply extend the context manager to control how much you would like to mix between the two seeds (if you want to):
```python
rng = dgl.random.cooperative_rng([seed1, seed2], ratio=[0.2, 0.8])
with rng:
    sg = dgl.sampling.sample_labors(g, seeds, fanout, ...)
```

mfbalin replied:
These are actually part of the labor sampling algorithm, no other sampling algorithm can do such things as far as I know. However, they are not documented anywhere yet, but will be soon.

I would prefer if it would get in with this functionality as the code review takes time and by the time this code is ready to be merged, I will be closer to completing an arxiv paper on cooperative minibatching. If an explanation is required, I could write an informal post to document this functionality and its benefits and experimental results, and the docstrings can point to the blog so that everything is documented and explained properly. (I already have some preliminary results)

If the above doesn't sound good to you, then since I want the PR to move forward, I will remove the random seeds and batch and layer dependency arguments. In the future, I will open another PR asking to change the interface from top to bottom. Please let me know what you think.

BarclayII replied:
> If an explanation is required, I could write an informal post to document this functionality and its benefits and experimental results, and the docstrings can point to the post so that everything is documented and explained properly.

That will be awesome.  I would imagine that it can justify some of the design/implementation decisions (e.g. the necessity of having individual RNG for each vertex, the necessity of using PCG), etc.  An informal post is enough - we just need some place to hold the information on why we are implementing the algorithm in these particular ways.

mfbalin replied:
For now, I will hold off on having the multiple random seeds in this PR. For the rest of the design decisions in this PR, I can still write a README in the LABOR sampling example PR. The example PR can also maybe include all of the possible things that LABOR can do, along with NIDs and random seed tricks with code snippets. Since you asked me whether CPU parallelism is possible, it is actually possible to parallelize labor sampling from the python side. I can also explain how this can be done in the README in the example PR.

BarclayII replied:
I would prefer to describe the design decision in this PR, since (1) it's harder to lose track, and (2) it's actually different from what LABOR sampling can do; it's more about "why we are implementing LABOR sampling this way".  I don't think the example PR is the best place for having them.  What do you think?

mfbalin replied:
Okay then, I will update the docstrings to be as explanative as I can. Will put a few more commends in the C++ implementation explaining how things work.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,276 @@
+    tensor([False, False, False])
+
+    """
+    if F.device_type(g.device) == 'cpu' and not g.is_pinned():
+        frontier, importances = _sample_labors(
+            g, nodes, fanout, random_seed, edge_dir=edge_dir, prob=prob,
+            importance_sampling=importance_sampling, copy_ndata=copy_ndata,
+            copy_edata=copy_edata, exclude_edges=exclude_edges)
+    else:
+        frontier, importances = _sample_labors(
+            g, nodes, fanout, random_seed, edge_dir=edge_dir, prob=prob,
+            importance_sampling=importance_sampling, copy_ndata=copy_ndata, copy_edata=copy_edata)
+        if exclude_edges is not None:
+            eid_excluder = EidExcluder(exclude_edges)
+            if importance_sampling != 0:
+                frontier, importances = eid_excluder(frontier, importances)
+            else:
+                frontier = eid_excluder(frontier)
```
I think this logic is common with `sample_neighbors`.  Is it possible to have a utility function to wrap this up?

mfbalin replied:
How can I handle the difference in the number of return returns values? Labor returns edge weights but sample_neighbors doesn't.

mfbalin replied:
Where could such a utility function be placed?

BarclayII replied:
`dgl/sampling/utils.py` could be one.

BarclayII replied:
Arguments can be handled by `(*args, **kwargs)`.  Return values is a problem though.  Let's table this for now.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,276 @@
+            "Please pass in a dictionary with empty tensors as values instead.")
+    ctx = utils.to_dgl_context(F.context(next(iter(nodes.values()))))
+    nodes_all_types = []
+    NIDs_all_types = []
+    for ntype in g.ntypes:
+        if ntype in nodes:
+            nodes_all_types.append(F.to_dgl_nd(nodes[ntype]))
+            try:
+                if len(g.ntypes) > 1:
+                    NIDs_all_types.append(F.to_dgl_nd(g.ndata[NID][ntype]))
+                else:
+                    NIDs_all_types.append(F.to_dgl_nd(g.ndata[NID]))
+            except:
+                NIDs_all_types.append(nd.array([], ctx=ctx))
+        else:
+            nodes_all_types.append(nd.array([], ctx=ctx))
+            NIDs_all_types.append(nd.array([], ctx=ctx))
```
This seems to be reading the "original node IDs" of the graph.  Why is that necessary?

mfbalin replied:
This is so that if we are sampling from a local graph, then global id's are used so that different machines can roll same random numbers for the same vertices. Same reason as using a random seed. Open to another suggestion to handle this.

mfbalin replied:
There can be an option to disable this so that there is less overhead in terms of reading from an extra array.

BarclayII replied:
> This is so that if we are sampling from a local graph, then global id's are used so that different machines can roll same random numbers for the same vertices.

I don't quite understand here.  By saying "different machines roll the same random numbers for the same vertices", you imply that the graph is replicated across all the machines?  If so, the "original node IDs" would not be necessary?

mfbalin replied:
The graph doesn't have to be replicated, it can be partitioned also. Cooperative minibatching PR partitions a graph to different GPUs and can run LABOR sampling on multiple GPUs with same outcome as if it was run on single machine with the same outcome.

mfbalin replied:
When one partitions the graph and also renames the vertices locally, then this functionality is needed to roll same random numbers for same vertices with different names in different partitions.

BarclayII replied:
So the necessity of NID is also a part of the cooperative minibatching?  My general opinion is that the logic of cooperative minibatching should go to the cooperative minibatching PR (or another PR about cooperative minibatching on LABOR sampling).

mfbalin replied:
Rolling a single random number for a vertex is crucial for labor to work for the scenarios vertices are renamed, for example when one is sampling from multiple subgraphs when the user passes relabel_nodes=True. If a batch were to consist of sampled subgraphs from different such subgraphs, then this NID logic is needed for labor sampling to work globally. Otherwise, its scope would be limited to each of those subgraphs, and it would sample more than it could have without this option.

BarclayII replied:
> If a batch were to consist of sampled subgraphs from different such subgraphs, then this NID logic is needed for labor sampling to work globally.

Is this a common use case for LABOR sampling (without considering cooperative minibatching)?

mfbalin replied:
I haven't figured out all the possible use cases yet. There may be any number of reasons a user might want to have such a sampling scheme. Since this is a new algorithm, I don't know how the users might want to take advantage of such a functionality but this logic is needed if we want the users to have this kind of flexibility.

Renaming the vertices of a subgraph from the original graph has the advantage that all the vertex ids become smaller. This might make the overall code faster also, for example radix sorting in cuda needs more passes to sort, the larger the vertex ids. Operations with the subgraph such as SpMM also become faster since accesses are more closely located in memory.

The addition of this feature harms the users of DGL in no way, but brings them flexibility in what additional things they can do. If you insist, I can add an option to use the NIDs, which could be turned off by default, and the users would have to manually opt-in if they want NIDs to be used.

BarclayII replied:
If we don't have an existing use case, we risk over-designing the API, and it complicates the implementation and maintenance of the code in the future (particularly by other people).  We (the DGL team) made the same mistakes before and it took a lot of effort to refactor out the unnecessary bits and further optimize the codebase.  So now we are trying to justify every design decision.

In this case, I think you can keep it as long as you justify the necessity with actual use cases (cooperative minibatching can be one for example).  It probably goes in the effort of having the informal post on the design decisions.

BarclayII replied:
For the NID argument, I think we can remove it from the Python API since the cooperative minibatching user-facing API is not settled.  We can still keep it in C++ interface since it's related to your future cooperative minibatching PR (with a comment describing why we need it).

mfbalin replied:
Disabled passing NIDs to the C++ side as per your suggestion along with a comment on what it was useful for.

mfbalin replied:
The default argument for the nids in C++ side is NullArray() and it is the last argument in the API, so the user doesn't even have to pass anything for the nids argument in the C++ api.

mfbalin replied:
I have done the same thing with the random seed, it is the second to last argument, and passing nothing for random_seed in the c++ API makes the code use a random seed that it gets from the DGL PRNG. Python side does the same thing, calling set_seed with no arguments sets the seed to a number it gets from the DGL PRNG.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,276 @@
+    if exclude_edges is not None:
+        if not isinstance(exclude_edges, dict):
+            if len(g.etypes) > 1:
+                raise DGLError("Must specify etype type when the graph is not homogeneous.")
```
"etype"

mfbalin replied:
I copied this code from python/dgl/sampling/neighbor.py, should I change that file too?

BarclayII replied:
Yes please.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,276 @@
+    Parameters
+    ----------
+    g : DGLGraph
+        The graph.  Can be either on CPU or GPU.
```
Does it work for heterogeneous graph?  Better document it if so.

mfbalin replied:
It does, just like neighbor sampler

BarclayII reviewed (changes requested):
High level comments in the interface design and docstrings.  Will dive into the C++ implementation later.

yaox12 commented:
@dgl-bot

yaox12 commented:
@dgl-bot

yaox12 commented:
@dgl-bot

yaox12 commented:
Generally LGTM.

yaox12 commented:
@dgl-bot

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,341 @@
+    Returns
+    -------
+    DGLGraph
+        A sampled subgraph containing only the sampled neighboring edges.
```
Missing the docstring on importance weights.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,341 @@
+    >>> g = dgl.graph(([0, 0, 1, 1, 2, 2], [1, 2, 0, 1, 2, 0]))
+    >>> g_edges = g.all_edges(form='all')``
+    (tensor([0, 0, 1, 1, 2, 2]), tensor([1, 2, 0, 1, 2, 0]), tensor([0, 1, 2, 3, 4, 5]))
+    >>> sg = dgl.sampling.sample_labors(g, [0, 1], 3, exclude_edges=[0, 1, 2])
```
This doesn't seem right.

mfbalin replied:
I run the code and that is the output. replacing sample_labors with sample_neighbors gives the exact same output.

BarclayII replied:
Oh never mind I misread.

BarclayII commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,341 @@
+        )
+        if exclude_edges is not None:
+            eid_excluder = EidExcluder(exclude_edges)
+            if importance_sampling != 0:
```
Isn't `importance_sampling` a `bool`?  Also, it seems that if I specify `exclude_edges` with `importance_sampling = False`, the length of importance array will be different from the number of edges in `frontier`.  Is that the correct behavior?

mfbalin replied:
importance sampling is an integer, 0 disables it, any other value enables it. Negative values indicate LABOR-*, nonnegative value i indicate LABOR-i. When importance_sampling is 0, no edge weights are returned unless weighted sampling is performed. In the weighted sampling case, there is a bug here, which I will fix now.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,346 @@
+    CSRMatrix mat,
+    IdArray NIDs,
+    IdArray rows,
+    int64_t __num_picks,
```
Why double underscore?

mfbalin replied:
I wanted to change its type to IdType with name num_picks below, that is why I gave it a temp name.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+    IdArray cnt,
+    int importance_sampling) {
+  using namespace aten;
+  const IdxType* indptr = static_cast<IdxType*>(mat.indptr->data);
```
You can use `mat.indptr.Ptr<IdxType>()`.  Same for other occurrences.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+  const IdxType* rows_data = static_cast<IdxType*>(rows->data);
+  const IdxType* nids = NIDs->shape[0] ? static_cast<IdxType*>(NIDs->data) : nullptr;
+  const auto num_rows = rows->shape[0];
+  const IdxType num_picks = __num_picks;
```
Any problem with using `int64_t`?

BarclayII replied:
No I mean why not use `int64_t` instead of `IdxType` for `num_picks`.

mfbalin replied:
Because num_picks is compared and used with variables of type IdxType, so when I call a function like std::min, then if both arguments are of type IdxType, no casting is needed.

mfbalin replied:
But it turns out that kind of thing happens on a single line so I will just use int64_t

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+    const IdxType rid = rows_data[i];
+    const auto act_d = indptr[rid + 1] - indptr[rid];
+    max_d = std::max(act_d, max_d);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_d;
```
Do a sanity check on the weight values (`d > 0`)?

mfbalin replied:
If d == 0, then accumulate is run with an empty range, then it will return 0.

BarclayII replied:
But in L94 you are dividing `num_picks` by `d`.  Is that OK though?

mfbalin replied:
d is a double so it is.

mfbalin replied:
If d is zero, then that value is not used at all.

mfbalin replied:
I have run the code will different datasets and it didn't crash even once over hundreds of training runs.

BarclayII replied:
I see.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  FloatArray cs_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
```
You can use `DGLDataTypeTraits<FloatType>::dtype`.  Same for other occurrences.  Even better, you can just define a variable `idtype` for this.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+        }
+        FloatType c = cs[i];
+        double var_1;
+        if (linear_c_convergence) {
```
If `linear_c_convergence` is `constexpr false`, why do we need the if-else block?

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+            hop_map[it.first] *= it.second;
+      }
+
+      for (int64_t i = 0; i < num_rows; ++i) {
```
Can we parallelize this loop?  If so, you can use `runtime::parallel_for`.

mfbalin replied:
If a parallel implementation is important, then I can implement the hashtable parts in other ways to make it parallel also. I could do that later also if needed.

mfbalin replied:
Also, if the cpu implementation needs to be improved, then one needs to implement a vectorized PRNG so that same logic in the cuda implementation can be used without the use of a hash_map. PCG was slow so I had to reduce the number of calls to it by using a hash_map to cache the computed random numbers.

BarclayII replied:
> Also, if serial parts aren't parallelized, then it will be a very huge bottleneck, and nobody should use it like that.

So you are suggesting that this for loop is actually computationally lighter than the other parts?  If that's the case, I think you can put it as a comment.

> Maybe I can make a proper parallel CPU implementation in a later PR? I think it will require a huge change to the code.

You don't have to.

> Also, if the cpu implementation needs to be improved, then one needs to implement a faster PRNG with same capability so that same logic in the cuda implementation can be used without the use of a hash_map. PCG was slow so I had to reduce the number of calls to it by using a hash_map to cache the computed random numbers.

I have a separate question on why to use PCG instead of STL's RNG but we can move our discussion there.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+  FloatArray cs_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* cs = static_cast<FloatType*>(cs_array->data);
```
I saw very abbreviated variable names like `cs`, `ds`, `ps`, `A`, `var_1`, `w`, `r`, etc.  Are they consistent with the notations in the paper?  If so, you can put a comment before the function signature referencing the paper.  If not, try naming them with the actual meaning.  Alternatively, you can leave a comment in the declaration of each variable to the notations in the paper (if they are different).

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+                                        ctx);
+  FloatType* ds = static_cast<FloatType*>(ds_array->data);
+
+  IdxType max_d = 1;
```
`max_degree`.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_d = indptr[rid + 1] - indptr[rid];
```
`actual_degree`?

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+
+    phmap::flat_hash_map<IdxType, FloatType> hop_map2;
+    for (int iters = 0; iters < importance_sampling || importance_sampling < 0; iters++) {
+      if (!weights || iters) {
```
I can't quite map the logic of this if-block to the algorithm.  Could you describe what it is doing as a comment?  (Or you can give a reference to the paper).

mfbalin replied:
There is no reference in the paper. It is an implementation detail about when there are no weights, then the first c values can be computed in O(1) because all the edge weights are uniform. The c value is just k / d where k is fanout and d is the degree. But when there are weights, since the computed c values are incorrect, I have to skip some logic making use of it to compute the right c values below.

mfbalin replied:
The later iterations will have correct c values so they will take the if in that case.

mfbalin replied:
I also increase importance_sampling by 1 above so that the weighted case can do one more of this iteration to match the unweighted case.

mfbalin replied:
I will improve the overall logic and seperate things into functions so that the code becomes more readable.

mfbalin commented:
@BarclayII, 2 random seeds are used when one wants to use batch_dependency, making consecutive batches dependent on each other by smoothly switching between 2 seeds over batch_dependency batches. The other 2 numbers keep track of how many batches were generated and the batch dependency information.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+
+  return std::make_pair(COOMatrix(mat.num_rows, mat.num_cols,
+                   picked_row, picked_col, picked_idx), importances);
+}
```
Could we break up this function into smaller functions?  For instance, I feel the `if (importance_sampling)` block can be put into a separate function, and L211-258 can be put into another, L279-309 feels like another.

A rule of thumb is to have each function do one thing and do it well.  If a function is doing many things at once, break it into smaller functions.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+    }
+  }
+
+  assert((IdxType)off == num_edges);
```
Is this a bug check?  Use `CHECK` with clear message as `assert` will crash the program with `SIGABRT` which is not friendly.

BarclayII reviewed (changes requested):
My high-level feeling on the CPU implementations.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+    for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+      const auto v = indices[j];
+      const auto u = nids ? nids[v] : v;
+      auto itb = rand_map.emplace(u, 0);
```
For instance, the meaning of `itb` doesn't look clear to me.  `it` can be translated to "iterator", so what does `b` mean?

mfbalin replied:
emplace returns a pair, an iterator to the inserted and if insertion was unsuccessful because the key already exists, a boolean to indicate that.

BarclayII replied:
I see.  You can leave it like this.

mfbalin replied:
With C++17, one can write `auto [it, b] =` instead of `auto itb =`, then use it and b as variables, but I reverted that change because windows build broke and dgl code base is officially c++14.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+  const auto a0 = std::cos(pi * l / 2);
+  const auto a1 = std::sin(pi * l / 2);
+
+  const pcg32 ng0(seeds[0]), ng1(seeds[num_seeds - 1]);
```
Why use `pcg32` instead of STL's own?

mfbalin replied:
Because I need a PRNG that can be used as a function f(seed, vertex_id). STL's PRNGs can not be used like this. What I do is I seed the PRNG with seed and skip to the vertex_id'th number to make it a function of seed and vertex_id.

mfbalin replied:
This ensures that with same seed and same vertex id, any machine rolls the same random number, making LABOR sampling distributed with 0 communication.

mfbalin replied:
Not to mention, none of the STL PRNGs are of high quality, I found this out from doing a literature search on this topic before implementing this algorithm, I wouldn't recommend anybody to use them if high quality randomness is needed. They fail many statistical tests, which I need for this algorithm to work properly.

BarclayII replied:
> Because I need a PRNG that can be used as a function f(seed, vertex_id) and for different values of seed and vertex_id, rolled randoms need to be uncorrelated, STL doesn't satisfy this requirement. What I do is I seed the PRNG with seed and skip to the vertex_id'th number to make it a function of seed and vertex_id.

A function of seed and vertex_id can also be as simple as like `linear_congruent(seed) ^ vertex_id`.  So would the correlation impact the model performance noticeably?  If not, I feel like we are adding additional technical complexity to the implementation with only theoretical benefit.

> Not to mention, none of the STL PRNGs are of high quality, I found this out from doing a literature search on this topic before implementing this algorithm, I wouldn't recommend anybody to use them if high quality randomness is needed. They fail many statistical tests, which I need for this algorithm to work properly.

Did you observe a gain from using PCG comparing to STL?  If so, by how much?  I'm not quite convinced by introducing a new PRNG solely because of better statistical randomness.

mfbalin replied:
I tried doing it your way while I was first developing labor by passing a seed as seed ^ id to STL PRNGs and sampling from it. The results degrade so badly that it looked broken with all numbers significantly different than the cuda implementation.

mfbalin replied:
But what I can do now that we don't have the random seed logic, is that I can remove the pcg32 and either use DGL's PRNG or an STL PRNG.

BarclayII replied:
> I tried doing it your way while I was first developing labor by passing a seed as seed ^ id to STL PRNGs and sampling from it. The results degrade so badly that it looked broken with all numbers significantly different than the cuda implementation.

This is very intriguing and definitely worth attention.   Do you still have the numbers?  And is the degradation happening on single-machine without cooperative minibatching as well?

I'm not entirely objecting using PCG, especially if not using it will cause a serious breakdown and no longer matches the performance in the paper.  But since it introduces an extra dependency and it requires specific handling (i.e. a hashmap for caching the results), I think it's better to leave a comment in code describing (1) why you prefer PCG over STL (i.e. the performance degradation, preferably with numbers), and (2) what's the issue of using PCG (i.e. slowness, you have to use a hashmap, etc.).

BarclayII replied:
The same question goes for having two RNGs with different seeds.  I feel it also needs explanation since it's a non-trivial decision.

mfbalin replied:
Sorry for the confusion, the current state of the code has no quality compromise. Using pcg, I could avoid using a hashmap using the same trick in the cuda side but since pcg was slow, I had to reduce the number of calls to it and use a hashmap instead. So using pcg, the code didn't speed up unlike the cuda case. Then, pcg was only needed to get the cooperative minibatching functionality for the cpu side. So I removed it.

The cuda side uses the prng as if it is a function of vertex id and the seed and calls the prng for each edge. Pcg was slow, so I had to cache the random results belonging to particular vertex ids in a hashmap. But if a hashmap is going to be used, then I can use DGL's PRNG too.

mfbalin replied:
To clarify, pcg is also needed for heterogenous graphs so that sampling is done globally for the graph with each edge type having the same random seed. Without pcg, this is not possible on the CPU side.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,347 @@
+          ng.discard(u);
+          norm.reset();
+          rnd += a1 * norm(ng);
+          itb.first->second = std::erfc(-rnd * (FloatType)/*M_SQRT1_2*/0.70710678118654752440) / 2;
```
What is this block doing?

mfbalin replied:
It interpolates between two random seeds smoothly so that batch_dependency works.

BarclayII commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+
+  // compute in-degrees
+  auto in_deg = allocator.alloc_unique<IdType>(num_rows + 1);
+  auto cs_arr = NewFloatArray(num_rows, ctx, sizeof(FloatType) * 8);
```
We can define a `floattype` variable and use `NDArray::Empty`.
If you prefer `NewFloatArray`, you can rewrite the `NDArray::Empty` implementation to use `NewFloatArray` instead.  Better make the code consistent.

mfbalin replied:
For IdArrays, there is a helper function NewIdArray, I wrote this to match the IdArray case.

BarclayII replied:
Got it.

BarclayII commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+  }
+
+  if (importance_sampling) {
+    { // extracts the onehop neighborhood cols to a contigous range into hop_1
```
contiguous

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+    const auto v = nids ? nids[u] : u;
+    curandStatePhilox4_32_10_t rng;
+    if (labor)
+      curand_init(123123, rand_seed, v, &rng);
```
there is no reason for 123123, it was randomly chosen by me. I read from the documentation that if one wants statistically uncorrelated randoms, then same random seed should be used. That is why I pass 123123 as the seed. This way of using it uses 123123 as the seed and samples the 2^64*rand_seed + v th number in the PRNG sequence. So, the rolled random number is a function of rand_seed, vertex id and also all the rolled randoms are statistically uncorrelated.

mfbalin replied:
With different seeds, given an integer i, ith number from two different seeds has the potential to be correlated.

mfbalin replied:
Rolling randoms that depend on the vertex id and seed is an implementation detail that can be taken advantage of to get other functionalities. It also eliminates a sort or the use of a hashmap, so it is an optimization, to ensure a single random is rolled for the same vertex.

This seed logic is what makes labor sampling efficient, otherwise I couldn't have shown that it is faster than neighbor sampling.

mfbalin replied:
if labor is set to false, then labor sampling reduces to neighbor sampling, will put a comment referencing the paper.

BarclayII replied:
Why would rolling for each vertex ID individually remove the necessity of a hashmap?

mfbalin replied:
Because the random to be rolled, called r_t in the paper where t is the vertex id has to be a function of the random seed and the vertex id t. If I don't use this trick, then I have to take the unique operation on the next layer's potential vertices, roll the random numbers and then scatter the rolled randoms back to the edges.

mfbalin replied:
If the rolled random was called r_{ts} so that the rolled random for each edge is different, where t is source id, s is destination id when sampling in edges, then we get the same behaviour as neighbor sampling. setting the constexpr labor to false gets this behaviour and then this code would turn into neighbor sampling.

mfbalin replied:
One can set the constexpr labor variable to false to see how much slower this implementation is compared to the cuda neighbor sampling implementation when doing the same work, which is doing neighbor sampling.

mfbalin replied:
If you checkout the current version of the cpu implementation, you will see that we store the rolled randoms into a hashmap. Then we lookup random numbers from the hashmap while doing sampling. This trick avoids that kind of extra computation/communication(meaning memory access).

mfbalin replied:
NVIDIA documentation says the following: Sequences generated with different seeds usually do not have statistically correlated values, but some choices of seeds may give statistically correlated sequences.

BarclayII replied:
> If I don't use this trick, then I have to take the unique operation on the next layer's potential vertices, roll the random numbers and then scatter the rolled randoms back to the edges which is how I was doing things before I noticed this trick.

Does that mean I no longer need a unique operation if each seed node gets its own RNG?  If two seed nodes share some neighbors, it's still possible for two vertices to pick the same neighbor no?

mfbalin replied:
It is not that each seed node gets its own RNG, let's say we are iterating over an edge with seed node s and neighbor t. Since we init the rng with random seed and t, the rolled random for this edge will be the same as the rolled random for any other edge with seed node s' and neighbor t.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+                  IdArray cnt,
+                  int importance_sampling) {
+  constexpr bool use_csr_spmv = false;
+  constexpr bool use_hashtable = false;
```
These were slower, but if the hashtable is improved in the future, this execution path might end up becoming faster. The CSR conversion overhead didn't bring enough benefits for spmv in the importance sampling iterations so it is disabled for now. But if the conversion was optimized in the future, this code path might end up becoming faster.

Should I remove them in this case?

BarclayII replied:
Yes.  You can leave a comment here with a link to your alternate implementation with `use_csr_spmv` and `use_hashtable` turned on so that we can revisit the idea when the optimization comes in.  Even better, you can give more details on why you decided turning them off (e.g. "performance is slower by 10% on dataset foobar" etc.)

Alternatively, you can define macros instead like
```c++
#ifdef LABOR_SAMPLING_USE_CSR_SPMV
...
#endif
```
These macros can then be controlled by CMake.  I would back this choice only if these paths are indeed noticeably faster on certain hardware so that they need to be turned on in some cases.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+                  int importance_sampling) {
+  constexpr bool use_csr_spmv = false;
+  constexpr bool use_hashtable = false;
+  const IdType num_picks = num_picks__;
```
Most other variables come in arrays of IdType, when I want to use std::min for example between a number of IdType and num_picks, then I would need an extra cast. This avoids those types of inconveniencies.

mfbalin replied:
num_picks can be at most as much as number of nodes in the graph. That is why it makes sense that it is of type IdType, otherwise I would need to keep casting it and treat it in a special way. CPU side was using it only once so I changed it but CUDA side seems to use it a lot, so I do the casting once and for all.

BarclayII replied:
OK.  I would prefer the argument name without the double underscores though, since they are more user-facing.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+namespace {
+
+template <typename ctx_t>
+class workspace_memory_alloc {
```
Sure, any code that makes use of thrust can make use of this class. There are also helper allocation routines that use smart pointers so that lifetime of the allocated pointers are managed automatically.

mfbalin replied:
Where should I place the new header file?

BarclayII replied:
Sounds good.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,858 @@
+  typedef char value_type;
+
+  void operator()(void * ptr) {
+    runtime::DeviceAPI::Get(*ctx)->FreeWorkspace(*ctx, ptr);
```
There was a compile error back when I was implementing it which I couldn't fix without using a pointer. Would appreciate the help if the actual object is need to be passed.

BarclayII replied:
Interesting.  You can leave a TODO in the comment and describe what the error is there.

mfbalin commented on src/graph/sampling/neighbor/neighbor.cc:
```diff
@@ -18,13 +22,15 @@ using namespace dgl::aten;
     const HeteroSubgraph& sg,
-    const std::vector<IdArray>& exclude_edges) {
+    const std::vector<IdArray>& exclude_edges,
+    const std::vector<FloatArray> *weights = nullptr) {
```
If exclusion is performed separately on edges and weights, then it will be slower. If I leave this function unchanged and write a version with weights, then there will be code duplication.

mfbalin replied:
Also, excluding edges after sampling is performed breaks the LABOR sampling algorithm. There ultimately needs to be a solution where the sampling is done from edges that can be included.

mfbalin replied:
There was a google doc about excluding edges etc. I proposed a solution with no overhead here: https://github.com/dmlc/dgl/issues/4441#issuecomment-1221634762

mfbalin replied:
Or one use weighted sampling for labor where one passes 0s for the edges that need to excluded. Weighted sampling should be faster than the existing cuda weighted neighbor sampling, didn't test them enough though.

BarclayII replied:
> If exclusion is performed separately on edges and weights, then it will be slower. If I leave this function unchanged and write a version with weights, then there will be code duplication.

Hmm I see.  You can leave this as a comment in the code then.

yaox12 commented:
@dgl-bot

BarclayII commented:
@dgl-bot

mfbalin commented:
@BarclayII, thanks a lot for the reviews. I have removed the pcg submodule, CPU side doesn't have the capability for cooperative minibatching anymore. If one manually sets the DGL seeds so that the CUDA side rolls the same numbers when getting the random seed, the cuda side can still be used with cooperative minibatching when running on multiple GPUs. The seed logic on the CUDA side is an optimization so there is no need to worry about it.

I have also removed unused code paths and addressed most of your reviews. I will work on refactoring the CPU and the CUDA implementations and try to improve readability by adding more comments. Meanwhile, if you can review all the remaining parts of the PR, I think the code will get into a shape that can be merged.

yaox12 commented:
@dgl-bot

yaox12 commented:
@dgl-bot

BarclayII commented:
@dgl-bot

mfbalin commented:
In these 2 PRs, I have removed the random seed logic, https://github.com/dmlc/dgl/pull/4668/commits/7f99034353080308f4783f27d9a08bea343fb796 and https://github.com/dmlc/dgl/pull/4668/commits/c250e07ac6d7e13f57e79e8a2c2f098d777378c2. If we come to a consensus on which of the removed parts I am allowed to revert as part of this PR, that would be great. For example NID logic is useful for cooperative minibatching, or if one relabels the vertices in the case of a heterogenous graph for example so that each relation type has their own local labels with NIDs being the global labels, then it would be needed. 2 random seeds is needed for batch dependency, so this PR doesn't really need multiple random seeds just yet. The overall random seed logic is needed both for cooperative minibatching and also for the heterogenous graphs. If the random seed logic is not there, then the rolled random numbers for different relation types will be different. I wouldn't want anybody run LABOR on heterogenous graphs without the random seed logic and compare it against their sampler.

yaox12 commented:
@dgl-bot

BarclayII commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,186 @@
+        self.set_seed()
+
+    def set_seed(self, random_seed=None):
+        '''updates the underlying seed for the sampler'''
```
This is a new method so a more concrete description is necessary (i.e. description of the parameters).  Also needs proper capitalization (i.e. `Updates`)

mfbalin replied:
I have updated this part with more details, along with the sample_labors documentation and also the C++ side documentation.

BarclayII commented:
@dgl-bot

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,324 @@
+          ps[j - indptr[rid]] = hop_map[indices[j]];
+      }
+
+      // stands for right handside of Equation (22) in arXiv:2210.13339
```
The RHS of Eq. 22 is $v_s$ so I guess the label is wrong.
However, I can see that this whole for loop is computing $c_s$ in Eq 17.  So maybe you can just put a comment above the for loop in L97 saying `compute c_s according to Eq.17`.

mfbalin replied:
In the extension for the weighted case, I left it as v_s, so that if a better way to choose v_s is found, then the improvement can be used. In the code, I set v_s as whatever I am computing there.

BarclayII replied:
I see.  So by saying "the RHS of Eq. 22", you actually mean "the RHS of Eq.22 after we move the other terms without `c_s` to RHS")?

mfbalin replied:
Yeah

BarclayII replied:
Then it's better to state it explicitly in the comment:
```C++
// stands for RHS of Equation (22) in arXiv:2210.13339 after moving the other terms without
// c_s to RHS.
```

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,324 @@
+          var_target += A[j] * A[j];
+      }
+      FloatType c = cs[i];
+      // stands for left handside of Equation (22) in arXiv:2210.13339
```
Same.

BarclayII replied:
This loop looks like it's doing fixed-point iteration to solve c, so I think the comment is incorrect.

mfbalin replied:
For both the cuda and cpu implementations, the O(nlogn) time computation described in the paper was slower. This implementation has complexity O(iters * n) time where iters is the number of iterations to converge. The larger n, usually the smaller number of iterations required. For most cases where fanout << n, it converges in a single iteration.

BarclayII replied:
Yes.  I'm just talking about this comment:
```
// stands for left handside of Equation (22) in arXiv:2210.13339
```
where it should be more like
```
// Compute c_s in Equation (22) via fixed-point iteration
```

mfbalin replied:
var_1 really stands for the lefthandside though.

mfbalin replied:
Though I could add an extra comment saying that fixed point iterations are utilized to compute c_s.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,324 @@
+
+      cs[i] = c;
+    }
+
+    if (!weights || iters) {
```
```suggestion
    // Check convergence.
    if (!weights || iters) {
```

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,324 @@
+      else
+        for (auto it : hop_map2)
+          hop_map[it.first] *= it.second;
+    }
```
I see.  So this is to compute the `hop_map` (in particular the necessary $\pi_s$, in other words Eq 18).  I will just say
```
Update the pi array according to Eq 18.
```
By the way, two questions:

Should Eq.18 in the paper be written as $\pi_t^{(i+1)} = \pi_t^{(i)} \max_{t \to s} c_s^{(i+1)}$?

Could we structure the code so that it looks like what you have said in the comment, i.e. if I understood correctly, something like
```C++
for (int iters = 0; iters < importance_sampling || importance_sampling < 0; ++iters) {
  for (int64_t i = 0; i < num_rows; ++i) {
    if (iters == 0 && !weights) {
      cs[i] = num_picks / d;
    } else {
      // Update cs[i] as usual...
    }
  }
  // Check convergence...
}
```
This way you probably don't need to add one to ``importance_sampling`` as well?

mfbalin replied:
But the last thing one needs to do is to recompute c_s with the updated pi values. If importance_sampling is 0, then I can avoid calling this function alltogether along with whatever it requires in terms of resources. So, I would either have to break out of the loop after updating c_s with your suggestion, and it would still require some logic. I think it is good as it is for now.

mfbalin replied:
The equation in the paper is correct.

BarclayII replied:
> The equation in the paper is correct.

Yeah I realized that $c_s(\pi^{(i)})$ refers to the solution to Eq. (15) given $\pi$ so you are right.  Never mind.

> But the last thing one needs to do is to recompute c_s with the updated pi values. If importance_sampling is 0, then I can avoid calling this function alltogether along with whatever it requires in terms of resources. So, I would either have to break out of the loop after updating c_s with your suggestion, and it would still require some logic. I think it is good as it is for now.

OK.  I'll try to rephrase the description in the code comment above then.

BarclayII commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,324 @@
+      // if hop_map is initialized, get ps from there, otherwise get it from the alternative.
+      num_edges +=
+          rnd <= (importance_sampling - weights ? c * hop_map[v] : c * (weights ? A[j] : 1));
+      rands[off++] = rnd;
```
Seems that you are precomputing the generated rolls.  Why not just generate them on-the-fly in L269?  This way you don't need to save an array, and the time complexity should be the same.

mfbalin replied:
I could generate them on the fly in the next loop also, but calls to the PRNG are expensive compared to array accesses. So this improves the performance slightly.

mfbalin replied:
I need to test it again to make sure this is faster though.

BarclayII replied:
Technically each element in `rands` array is written and read exactly once, so the number of PRNG calls should be the same.  So generating the numbers on the fly in L269 should save us from allocating `rands_arr` and writing to it.

However, if the current implementation is indeed noticeably faster, you can leave a comment in the code to tell people so that they (including myself) don't try to "optimize" it.  I can't think of a reason why the current implementation could be faster though (other than some caching happening around).

mfbalin replied:
Before, the calls to the PRNG were slower when I was using the skip ahead with the vertex id t which had O(log t) complexity. Now though, I decided to pass t as the stream argument, so it should be O(1). Now, not using an array could potentially be faster as I had made this optimization when PRNG calls were logarithmic complexity. Will test and get back here. Will resolve this conversation after changing to whatever is faster and putting a comment if needed.

mfbalin replied:
I have gotten rid of the two steps to compute the number of sampled edges and then allocating an array of that size. Doing it in a single step makes the code around 40% faster when importance_sampling=0 and there are no weights. Other scenarios should also be faster now.

BarclayII reviewed (changes requested):
I'll leave CUDA code for tomorrow.  In general I feel the code is much more readable now.

BarclayII commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,807 @@
+  auto rands = allocator.alloc_unique<FloatType>(importance_sampling ? hop_size : 1);
+  auto probs_found = allocator.alloc_unique<FloatType>(importance_sampling ? hop_size : 1);
+
+  if (weights) {  // the computed c values are wrong when there are weights, so we recompute.
```
This code organization (especially with the description saying "computed c values are wrong") is confusing.  Could we instead do
```C++
// compute c_s
if (weights) {
  // this block...
  const expr int BLOCK_CTAS = BLOCK_SIZE / CTA_SIZE;
  // ...
} else {
  // Move L596-L597 here
  // ...
}
```

mfbalin replied:
I don't want to launch a kernel just to compute c_s when there are no weights. It would make the performance worse.

mfbalin replied:
I opted for performance mostly, not code clarity. That is why some things may look out of place but I gave a lot of thought into how I can make things fast, expecially the importance_sampling == 0 case.

mfbalin replied:
My goal was to launch the fewest number of kernels possible.

mfbalin replied:
The kernel launch that inits cs does other things too, so I can't move it here.

BarclayII replied:
I see.  That makes sense.  To make the code clearer you can move the if block below L596-L597 so the code that does one thing (in this case computing `c_s`) are grouped together:
```C++
// If the graph is unweighted, we can compute c_s and in_deg in one kernel as c_s = k / in_deg.
// However, if the graph is weighted, we have to compute in_deg and c_s in separate kernels.
// L596-L597...
// L636-L655...
```

mfbalin commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,207 @@
+        
+        The passed random_seed makes it so that for any seed vertex s and its neighbor t, the rolled
+        random variate r_t is the same for any instance of this class with the same random seed.
+        When sampling as part of the same batch, one would want identical seeds so that LABOR
+        can globally sample. One example is that for heterogenous graphs, there is a single random
+        seed passed for each edge type. This will sample much fewer vertices compared to having
+        unique random seeds for each edge type. If one called this function individually for each
+        edge type for a heterogenous graph with different random seeds, then it would run LABOR
+        locally for each edge type, resulting into a larger number of vertices being sampled.
```
The reason why LABOR samples less than neighbor sampling in general is that it uses the same random seed for each seed vertex. If each seed vertex was to use a different random seed, then labor would turn into neighbor sampling. Since the whole idea and the reduction in the number of vertices sampled comes from using identical seeds, the same thing happens in the heterogenous case as well. I don't know how to further elaborate.

mfbalin commented on src/array/cpu/labor_pick.h:
```diff
@@ -0,0 +1,324 @@
+        hop_map);
+
+  const uint64_t random_seed = IsNullArray(random_seed_arr) ?
+      RandomEngine::ThreadLocal()->RandInt(1000000000) : random_seed_arr.Ptr<int64_t>()[0];
```
It is a tensor because the tensor can be empty in case the user doesn't pass anything. Then the C++ code can check if the tensor is empty and if it is, then can get the random seed from DGL PRNG. So, I am using the tensor like std::optional.

mfbalin replied:
And also, if the python side doesn't have that logic, then I can't ensure that the same random_seed is used for different edge types.

mfbalin replied:
This logic exists so that when anyone uses the c++ api, they don't have to pass anything for random seed, it will be handled automatically. Should improve c++ api user experience.

BarclayII replied:
OK.  You can keep it for now.

mfbalin commented on src/array/array.cc:
```diff
@@ -25,6 +25,10 @@ IdArray NewIdArray(int64_t length, DGLContext ctx, uint8_t nbits) {
   return IdArray::Empty({length}, DGLDataType{kDGLInt, nbits, 1}, ctx);
 }
 
+FloatArray NewFloatArray(int64_t length, DGLContext ctx, uint8_t nbits) {
```
Should I do the same for NewIdArray, I just followed the same practice for NewIdArray.

BarclayII replied:
Oh OK never mind.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,807 @@
+      // stands for k in in arXiv:2210.13339, i.e. fanout
+      const auto k = min(num_picks, degree);
+      // slightly better than NS
+      const FloatType d_ = ds ? ds[row] : degree;
```
The implementations are exactly the same. They would produce almost the exact same output if I had used the same PRNG for both. There is a slight logic difference between the CPU and CUDA implementations in terms of arrays used etc. but otherwise the code does the same thing. The CPU side ds array holds the sum of weights if there are weights, otherwise it holds the actual degree. The CUDA side doesn't have a valid ds array when there are no weights so there is a conditional there.

mfbalin commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,207 @@
+        importance sampling probabilities until convergence while use of positive values runs
+        optimization steps that many times. If the value is i, then LABOR-i variant is used.
+    layer_dependency : bool, default ``False``
+        Specifies whether different layers should use same random variates.
```
It is just an option that can reduce the number of vertices sampled at the expense of a slight reduction in quality. I updated the docstring.

mfbalin commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,207 @@
+        self.layer_dependency = layer_dependency
+        self.set_seed()
+
+    def set_seed(self, random_seed=None):
```
Just added one more sentence about when to call this function.

mfbalin commented on python/dgl/sampling/labor.py:
```diff
@@ -0,0 +1,356 @@
+        importance sampling probabilities until convergence while use of positive values runs
+        optimization steps that many times. If the value is i, then LABOR-i variant is used.
+    random_seed : tensor
+        The passed random_seed makes it so that for any seed vertex s and its neighbor t, the rolled
```
Added a sentence about when to set the seed.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,807 @@
+__global__ void _CSRRowWiseLayerSampleDegreeKernel(
+  const IdType num_picks,
+  const IdType num_rows,
+  const IdType * const rows,
```
Other cuda sampling implementations weren't using __restrict__ so I didn't use either.

BarclayII replied:
Hmm technically those buffers should use `__restrict__` as well.  @yaox12 @nv-dlasalle what do you think?

BarclayII replied:
We don't have to do this right now though.

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,807 @@
+  auto hop_new = static_cast<IdType*>(hop_new_arr->data);
+  auto hop_unique = allocator.alloc_unique<IdType>(hop_size);
+  // After this block, hop_unique holds the unique set of one-hop neighborhood
+  // and hop_new holds the renamed hop_1, idx_coo already holds renamed destination.
```
Just updated

BarclayII commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,807 @@
+    const auto out_row = rows[tIdx];
+    const auto d = indptr[out_row + 1] - indptr[out_row];
+    in_deg[tIdx] = d;
+    cs[tIdx] = num_picks / (ds ? ds[tIdx] : (FloatType)d);
```
Is `ds` a necessary argument for this operator?  If `ds` is not `nullptr`, it means the graph is weighted, and `cs` needs to be recomputed anyway.  Did I miss something?

BarclayII replied:
I see.

BarclayII commented:
@dgl-bot

mfbalin commented on src/array/cuda/labor_sampling.cu:
```diff
@@ -0,0 +1,807 @@
+
+}  // namespace
+
+template <typename IdType, typename FloatType, typename exec_policy_t>
```
I did, could you see if the explanations are enough?

BarclayII commented:
@dgl-bot

BarclayII commented:
@dgl-bot

BarclayII commented:
@dgl-bot

frozenbugs commented on include/dgl/aten/array_ops.h:
```diff
@@ -47,7 +47,18 @@ IdArray NewIdArray(
     uint8_t nbits = 64);
 
-/**
+/*!
```
/**

frozenbugs commented on include/dgl/aten/array_ops.h:
```diff
@@ -47,7 +47,18 @@ IdArray NewIdArray(
+                   DGLContext ctx = DGLContext{kDGLCPU, 0},
+                   uint8_t nbits = 32);
+
+/*!
```
ditto.

frozenbugs commented on include/dgl/aten/coo.h:
```diff
@@ -379,9 +379,64 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries);
-/**
- * @brief Randomly select a fixed number of non-zero entries along each given
- * row independently.
+/*!
```
ditto.

frozenbugs commented on include/dgl/aten/coo.h:
```diff
@@ -379,9 +379,64 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries);
+ * The picked indices are returned in the form of a COO matrix.
+ * 
+ * The passed random_seed makes it so that for any seed vertex s and its neighbor t, the rolled
+ * random variate r_t is the same for any call to this function with the same random seed.
```
80 char max.

frozenbugs commented on include/dgl/aten/coo.h:
```diff
@@ -379,9 +379,64 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries);
+ * // sampled.cols = [1, 2, 3]
+ * // sampled.data = [3, 0, 4]
+ *
+ * \param mat Input coo matrix.
```
\@param

frozenbugs commented on include/dgl/aten/coo.h:
```diff
@@ -379,9 +379,64 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries);
+ * \param importance_sampling Whether to enable importance sampling
+ * \param random_seed The random seed for the sampler
+ * \param NIDs global nids if sampling from a subgraph
+ * \return A pair of COOMatrix storing the picked row and col indices and edge weights
```
\@return

frozenbugs commented on include/dgl/aten/coo.h:
```diff
@@ -379,9 +379,64 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries);
+    IdArray random_seed = NullArray(),
+    IdArray NIDs = NullArray());
+
+/*!
```
ditto

frozenbugs commented on include/dgl/aten/coo.h:
```diff
@@ -379,9 +379,64 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries);
+    IdArray NIDs = NullArray());
+
+/*!
+ * \brief Randomly select a fixed number of non-zero entries along each given row independently.
```
ditto

frozenbugs commented on include/dgl/aten/csr.h:
```diff
@@ -407,9 +407,64 @@ CSRMatrix CSRReorder(
-/**
- * @brief Randomly select a fixed number of non-zero entries along each given
- * row independently.
+/*!
```
ditto and other places.

frozenbugs commented on include/dgl/aten/csr.h:
```diff
@@ -407,9 +407,64 @@ CSRMatrix CSRReorder(
+ *
+ * The picked indices are returned in the form of a COO matrix.
+ * 
+ * The passed random_seed makes it so that for any seed vertex s and its neighbor t, the rolled
```
max 80 chars, and others.

frozenbugs commented on python/dgl/dataloading/labor_sampler.py:
```diff
@@ -0,0 +1,211 @@
+    <https://arxiv.org/abs/2210.13339>`
+
+    This sampler will make every node gather messages from a fixed number of neighbors
+    per edge type. The neighbors are picked uniformly with default parameters. For every vertex t
```
max 80 chars.

BarclayII commented:
@dgl-bot

BarclayII reviewed (approved):
I'll approve this PR.  There are some more stuff left for code styles which was not enforced by lint and I think we can leave them for future PRs (possibly with the help from the script from @frozenbugs).

BarclayII commented:
@dgl-bot

mfbalin reviewed (commented):
@BarclayII Then, you might want to change the other occurrences of DGLHeteroGraph to DGLGraph in your last commit below also, unless I am missing something?

mfbalin commented:
Also, when all changes are final, can I run my example scripts to ensure that everything still works as expected? After that, it can be merged.

mfbalin commented:
> 

Everything seems to work. @BarclayII, @yaox12 and @frozenbugs thanks a lot for the reviews. So excited!

yaox12 commented:
Ready to merge?
